lib/AST/NameLookup.cpp

```diff
@@ -544,50 +596,47 @@ static void recordShadowedDecls(ArrayRef<ValueDecl *> decls,
         auto nominal = ctor->getDeclContext()->getSelfNominalTypeDecl();
         auto &knownInits = importedInitializerCollisions[nominal];
         if (knownInits.size() == 1) {
-          importedInitializerCollectionTypes.push_back(nominal);
+          importedInitializerCollisionTypes.push_back(nominal);
         }
         knownInits.push_back(ctor);
       }
     }
 
-    CanType signature;
+    // If the decl is currently being validated, this is likely a recursive
+    // reference and we'll want to skip ahead so as to avoid having its type
+    // attempt to desugar itself.
+    if (decl->isRecursiveValidation())
```

Comment:
Is this the right recursion guard? We may be able to scale back to `isComputingGenericSignature`.

Reply:
Well, we compute the interface type immediately below anyway. I kept the same guard as before because I didn't want to introduce a subtle behavioral difference here. I suspect we might be able to get away with removing these guards at some point, if not now. They're extremely unprincipled.